alpenhorn/queue.py

```diff
@@ -0,0 +1,426 @@
+            # will have as many elements as the maximum number of worker threads
+            # that have ever existed at one time, so it's probably not worth the
+            # trouble.
+
+            # Notify waiters when there are no pending tasks
+            if self._total_queued == 0 and self._total_inprogress == 0:
+                self._all_tasks_done.notify_all()  # wakes up all waiting threads
+
+    def join(self) -> None:
+        """Blocks until there's nothing left in the queue.
+
+        This includes waiting for in-progress tasks.  All deferred puts are
+        discarded, including ones added while this call is blocking.
+        """
+        # Discard deferred puts
+        with self._dlock:
+            self._joining = True
+            self._deferrals = list()
+
+        with self._all_tasks_done:
+            while self._total_inprogress > 0 or self._total_queued > 0:
+                self._all_tasks_done.wait()  # woken up by the notify_all() in task_done()
+
+        with self._dlock:
+            self._joining = False
+
+    @property
+    def qsize(self) -> int:
+        """Total number of queued tasks.
+
+        Excludes in-progress tasks and deferred puts not yet processed.
+
+        Not to be relied on."""
+        with self._lock:
+            return self._total_queued
+
+    @property
+    def inprogress_size(self) -> int:
+        """Total number of in-progress tasks.  Not to be relied on."""
+        with self._lock:
+            return self._total_inprogress
+
+    def fifo_size(self, key: Hashable) -> int:
+        """Total number of tasks queued or in progress for the FIFO named `key`.
+
+        Does not include deferred puts not yet processed.
+
+        Returns 0 for any non-existent FIFO (i.e for any `key` not previously used).
+
+        Parameters
+        ----------
+        key : hashable
+            The name of the FIFO to return the size of.
+
+        Returns
+        -------
+        fifo_size : int
+            The size of the FIFO as explained above.
+        """
+        with self._lock:
+            if key not in self._fifos:
+                return 0
+            return len(self._fifos[key]) + self._inprogress_counts[key]
+
+    @property
+    def deferred_size(self) -> int:
+        """Total number of deferred puts not yet processed."""
+        with self._dlock:
+            return len(self._deferrals)
+
+    def _new_fifo(self, key: Hashable) -> deque:
+        """Create and return a new FIFO named `key`.
+
+        A FIFO named `key` must not yet exist!
+
+        Adds `key` to both `_fifos` and `_inprogress_counts`.
+
+        The new FIFO is also added to the zero bin of
+        `_keys_by_inprogress`.
+
+        Parameters
+        ----------
+        key : hashable
+            The name of the FIFO to create.
+
+        Returns
+        -------
+        fifo : deque
+            The newly created FIFO
+        """
```

Comment:
Although this _is_ a private method, I'm kind of of the opinion that the logic for checking if a FIFO key already exists should be done here rather than in the `_put` method. A user could technically overwrite an existing queue by accessing this directly, and there wouldn't be any explicit warning about it. I'm also not sure if the logic of creating a new FIFO makes sense in the `_put` method (even though that is also private). 

My suggestion is to replace the `_new_fifo` method with something like a `_get_fifo` method which would return either an existing queue or create a new one as needed.

Reply:
I'm not really sure why this is separate.  What would you think about just merging the four operative lines down into the `_put` function at the only place where this is called?

Reply:
I think that would make sense, and it would remove the possibility of any issues happening (however unlikely). If the method isn't used anywhere else than it probably doesn't have to be its own method